### PR TITLE
fix(cpp): recover declaration-shaped stack constructions into ctor calls

### LINKED ISSUE #871: fix(cpp): stack-object construction misparsed as function declaration leaves ctor call trees dead
Found live-validating the Dart dogfood fixes on gskinnerTeam/flutter-wonderous-app (tree-sitter C++ frontend, no compile_commands.json). After #867 rooted `wWinMain`, 19 of the remaining 34 dead-code candidates are still `windows/runner` symbols, and the chain breaks at a classic most-vexing-parse ambiguity:

```cpp
FlutterWindow window(project);
```

tree-sitter-cpp parses this stack-object construction as a FUNCTION DECLARATION named `window` returning `FlutterWindow`. Consequences visible in the graph:

- A phantom `Function` node `windows.runner.main.window` is minted (and itself reported dead).
- No construction CALLS edge is emitted, so the `FlutterWindow`/`Win32Window` constructor chain never becomes reachable, and the header-declared method nodes (`win32_window.h.Win32Window.*`) stay an unreachable island: ctor/dtor, `MessageHandler`, `WndProc`, `GetHandle`, `Point.Point` all report dead.

The construction-redirect machinery from the ctor follow-up campaign (#791/#792/#797/#799/#800) handles temporaries and `new`, but a declaration-shaped construction never reaches it.

Proposed direction: recover the misparse at ingestion. A "function declaration" inside a function body whose declarator arguments are bare identifiers naming in-scope locals/params (not types) is a construction of the declared type: emit the local variable typing plus a construction CALLS redirect to the ctor, and do not mint a Function node for it. The declaration-vs-construction test can stay conservative (only fire when at least one argument resolves to a known local) so genuine nested declarations keep working.

Repro: index flutter-wonderous-app with `--capture all`, run `cgr dead-code`, observe the `windows.runner` block and the phantom `windows.runner.main.window` function.

### PR DESCRIPTION
**Pull request by vitali87**
## Summary

Fixes #871. `FlutterWindow window(project);` inside a function body is a stack-object construction, but tree-sitter-cpp parses it as a FUNCTION DECLARATION named `window` returning `FlutterWindow` (the classic most-vexing-parse ambiguity). cgr minted a phantom `Function` node for `window` and emitted no construction edges, so the whole `FlutterWindow`/`Win32Window` ctor chain stayed an unreachable island in dead-code.

Probing the repro also surfaced a second, adjacent gap: even a CORRECTLY parsed declaration construction (`Point origin(10, 10)`, an `init_declarator` with an argument list) emitted no ctor call. The construction-redirect machinery from the ctor campaign covers temporaries and `new`, but declaration-shaped constructions never reached it.

## Changes

- `cpp/utils.py`: `is_cpp_vexing_parse_construction` recognises the misparse conservatively. Every declarator "parameter" must be a bare type-less identifier and at least one must name an in-scope local or parameter of the enclosing function, so genuine nested prototypes (`int helper(int);`, `FlutterWindow maker(SomeType);`, and the standard's `FlutterWindow factory();`) keep working.
- `function_ingest.py`: a declaration passing that test mints no Function node (kills the phantom).
- `call_processor.py`: new per-caller pass `_ingest_cpp_declaration_ctor_calls`, alongside the member-init and braced-return passes. Both declaration shapes emit INSTANTIATES to the class plus CALLS to its ctors and the end-of-lifetime dtor via the existing `_emit_cpp_ctor_calls`. Revive-only: nothing is emitted unless the declared type resolves to a registered first-party class; nested lambda and local-class scopes are skipped.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

## Related Issues

Closes #871.

## Test Plan

RED first: five tests in `test_cpp_stack_construction.py`; four fail on main (phantom minted, no ctor/dtor CALLS, no INSTANTIATES, for both shapes), the prototype-conservatism guard already holds and pins it.

Full suite: 5891 passed. Live validation on flutter-wonderous-app: dead-code 21 -> 17; the phantom `windows.runner.main.window` is gone and `FlutterWindow.FlutterWindow`, `FlutterWindow.~FlutterWindow` and `Win32Window.Point.Point` cleared. The remaining windows/runner items are distinct gaps (implicit base-ctor calls, virtual dispatch through `WndProc`, header-prototype nodes) and will be triaged separately on the issue.

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved C++ ingestion to detect “most-vexing-parse” declaration-shaped stack-object constructions and recover constructor/destructor reachability when call nodes are missing.
  - Prevented misparsed constructs from creating phantom function entries.
  - Enhanced constructor call and instantiation edge recording for first-party C++ types.
  - Improved in-scope evidence matching across `if`/`switch` condition declarations and lambda scopes (including lambda default capture behavior), while keeping out-of-scope prototypes unchanged.
- **Tests**
  - Added a new integration-style C++ test suite covering these edge cases and expected call/instantiation graph outcomes.

### CODE REVIEW

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/cpp/utils.py [hunk 1]
```diff
@@ -538,3 +538,112 @@ def extract_class_name_from_out_of_class_method_qualified(
...
+        if child.type in cs.CPP_NESTED_SCOPE_NODE_TYPES:
+            continue
+        if child.type == cs.CppNodeType.PARAMETER_DECLARATION or (
+            child.type == cs.CppNodeType.DECLARATION and child is not node
+        ):
+            for declarator in child.children_by_field_name(cs.FIELD_DECLARATOR):
+                if name := _declarator_bound_name(declarator):
+                    names.add(name)
+            continue
+        stack.extend(child.children)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/cpp/utils.py]: Fixed in b84bb666 (RED test first: test_out_of_scope_names_keep_prototypes). The collector now walks the candidate's ancestor chain and counts only declarations that lexically PRECEDE it among each ancestor's direct children, plus the enclosing function's parameters; later declarations and sibling nested blocks are invisible, and a for-init declaration stays visible to a candidate in the loop body.

File: codebase_rag/parsers/cpp/utils.py [hunk 2]
```diff
@@ -538,3 +538,112 @@ def extract_class_name_from_out_of_class_method_qualified(
...
+    # parses this stack-object construction as a function DECLARATION named
+    # `window` returning FlutterWindow (issue #871). Conservative recovery:
+    # every declarator argument must be a bare type-less identifier and at
+    # least one must name an in-scope local or parameter, so genuine nested
+    # prototypes keep working.
+    names = cpp_vexing_parse_argument_names(decl_node)
+    if not names:
+        return False
+    in_scope = cpp_enclosing_function_value_names(decl_node)
+    return any(name in in_scope for name in names)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/cpp/utils.py:727]: Real bug, fixed in b84bb666 with a RED test covering exactly this shape (later same-scope local named like the parameter type). Evidence collection is now lexically scoped: preceding-siblings-only along the ancestor chain, parameters from the function declarator.

[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/cpp/utils.py:727]: Fixed in b84bb666; the RED test also covers the sibling nested block variant (a block-local declared after the prototype no longer counts as evidence). A candidate inside a lambda now takes evidence only from the lambda's own scope rather than the outer function.

**Comment by vitali87** [comment]
@greptile review

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/cpp/utils.py [hunk 3]
```diff
@@ -538,3 +538,147 @@ def extract_class_name_from_out_of_class_method_qualified(
...
+    # Declared names among `scope`'s direct children that lexically
+    # precede `node`.
+    for sibling in scope.children:
+        if sibling.start_byte >= node.start_byte:
+            return
+        if sibling.type != cs.CppNodeType.DECLARATION:
+            continue
+        for declarator in sibling.children_by_field_name(cs.FIELD_DECLARATOR):
+            if name := _declarator_bound_name(declarator):
+                names.add(name)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/cpp/utils.py:660]: Half refuted, half fixed in 3d7f6e85. The for case already worked: a for-init declaration IS a direct child of for_statement, so the ancestor-chain scan collects it (verified on this branch: `for (int project = 0; ...) { FlutterWindow window(project); }` emits the ctor call and mints no phantom; the RED run only failed on a branch without this PR). The if/switch case was real: the condition declaration nests one level down inside condition_clause, so the scan now descends into a preceding condition_clause sibling, with a RED test covering `if (int cproject = 1) { FlutterWindow cwindow(cproject); }`.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/cpp/utils.py [hunk 4]
```diff
@@ -538,3 +538,152 @@ def extract_class_name_from_out_of_class_method_qualified(
...
+    current = node.parent
+    while current is not None:
+        _collect_preceding_declaration_names(current, node, names)
+        if current.type == cs.CppNodeType.FUNCTION_DEFINITION:
+            _collect_cpp_parameter_names(current, names)
+            return names
+        if current.type in cs.CPP_NESTED_SCOPE_NODE_TYPES:
+            # A lambda or local-class member body opens its own scope; a
+            # candidate inside one takes evidence only from that scope.
+            return names
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/cpp/utils.py]: Fixed in 228a8e4d (RED test first: test_lambda_parameter_counts_as_evidence). Two parts: the evidence walk now collects a lambda's parameters when the candidate sits in its body (the abstract_function_declarator shape), and the declaration pass walks lambda bodies and declaration initializers, matching how the general call pass flat-attributes lambda-body calls to the enclosing caller.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/cpp/utils.py [hunk 5]
```diff
@@ -538,3 +538,160 @@ def extract_class_name_from_out_of_class_method_qualified(
...
+        _collect_preceding_declaration_names(current, node, names)
+        if current.type == cs.CppNodeType.FUNCTION_DEFINITION:
+            _collect_cpp_parameter_names(current, names)
+            return names
+        if current.type in cs.CPP_NESTED_SCOPE_NODE_TYPES:
+            # A lambda or local-class member body opens its own scope; a
+            # candidate inside one takes evidence only from that scope,
+            # which for a lambda includes its parameters.
+            if current.type == cs.TS_CPP_LAMBDA_EXPRESSION:
+                _collect_cpp_parameter_names(current, names)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/cpp/utils.py]: Fixed in 7f2c8287 (RED test first: test_lambda_capture_counts_as_evidence). Explicit capture identifiers now join the evidence set, and a default capture (`[=]`/`[&]`) makes the walk continue into the enclosing scope, since it pulls every enclosing local into view.

**Comment by vitali87** [comment]
@greptile review

File: codebase_rag/parsers/cpp/utils.py [hunk 6]
```diff
@@ -538,3 +538,180 @@ def extract_class_name_from_out_of_class_method_qualified(
...
+    # Collect explicit capture names; True when a default capture
+    # (`[=]` / `[&]`) is present, meaning every enclosing local is visible.
+    has_default = False
+    for child in lambda_node.children:
+        if child.type != cs.TS_CPP_LAMBDA_CAPTURE_SPECIFIER:
+            continue
+        for cap in child.children:
+            if cap.type == cs.CppNodeType.IDENTIFIER:
+                if name := safe_decode_text(cap):
+                    names.add(name)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/cpp/utils.py:673]: Fixed (RED test first: test_lambda_init_capture_counts_as_evidence): a lambda_capture_initializer contributes its LEADING identifier, closing the capture family alongside explicit captures and the `[=]`/`[&]` defaults.

**Comment by vitali87** [comment]
@greptile review
